File: backend/src/main/java/de/garrafao/phitag/domain/dictionary/entry/DictionaryEntry.java
package de.garrafao.phitag.domain.dictionary.entry;

import java.util.List;

import javax.persistence.*;

import org.apache.commons.lang3.Validate;

import de.garrafao.phitag.domain.dictionary.dictionary.Dictionary;
import de.garrafao.phitag.domain.dictionary.sense.DictionaryEntrySense;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@Entity
@Table(name = "phitagdictionaryentry")
@Getter
@ToString
public class DictionaryEntry {

    @EmbeddedId
    private DictionaryEntryId id;

    @MapsId("dictionaryid")
    @ManyToOne
    @JoinColumns({
            @JoinColumn(name = "dname", referencedColumnName = "dname"),
            @JoinColumn(name = "uname", referencedColumnName = "uname")
    })
    private Dictionary dictionary;

    @Setter
    @Column(name = "headword", nullable = false)
    private String headword;

    @Setter
    @Column(name = "partofspeech")
    private String partofspeech;

    @OneToMany(mappedBy = "dictionaryentry")
    private List<DictionaryEntrySense> senses;

    DictionaryEntry() {
    }

    public DictionaryEntry(final Dictionary dictionary, final String headword, final String partOfSpeech) {
        Validate.notNull(dictionary, "dictionary must not be null");
        Validate.notEmpty(headword, "headword must not be null");

        this.id = new DictionaryEntryId(dictionary.getId());
        this.dictionary = dictionary;

        this.headword = headword;
        this.partofspeech = partOfSpeech;
    }

    @Override
    public boolean equals(final Object obj) {
        if (!(obj instanceof DictionaryEntry)) {
            return false;
        }

        DictionaryEntry other = (DictionaryEntry) obj;
        return this.id.equals(other.id);
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

}

File: frontend/lib/model/dictionary/entry/model/DictionaryEntry.ts
import DictionaryEntrySense from "../../sense/model/DictionaryEntrySense";
import DictionaryEntryDto from "../dto/DictionaryEntryDto";
import DictionaryEntryId from "./DictionaryEntryId";

export default class DictionaryEntry {

    readonly id: DictionaryEntryId;

    headword: string;
    partofspeech: string;

    readonly senses: Array<DictionaryEntrySense>;

    constructor(id: DictionaryEntryId, headword: string, partofspeech: string, senses: Array<DictionaryEntrySense>) {
        this.id = id;
        this.headword = headword;
        this.partofspeech = partofspeech;
        this.senses = senses;
    }

    public copy(): DictionaryEntry {
        return new DictionaryEntry(
            this.id.copy(),
            this.headword,
            this.partofspeech,
            this.senses.map(sense => sense.copy())
        );
    }

    public shallowAnnonymizedCopy(): {
        id: {
            id: string,
            dname: string,
            uname: string,
        },
        headword: string,
        partofspeech: string,
    } {
        return {
            id: this.id.shallowAnnonymizedCopy(),
            headword: this.headword,
            partofspeech: this.partofspeech,
        };
    }

    public static fromDto(dto: DictionaryEntryDto) {
        return new DictionaryEntry(
            DictionaryEntryId.fromDto(dto.id),
            dto.headword,
            dto.partofspeech,
            dto.senses.map(DictionaryEntrySense.fromDto)
        );
    }

}